impeller/base/job_queue.h

```diff
@@ -0,0 +1,107 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_IMPELLER_BASE_JOB_QUEUE_H_
+#define FLUTTER_IMPELLER_BASE_JOB_QUEUE_H_
+
+#include <deque>
+#include <map>
+#include <memory>
+
+#include "flutter/fml/closure.h"
+#include "flutter/fml/concurrent_message_loop.h"
+#include "impeller/base/comparable.h"
+#include "impeller/base/thread.h"
+
+namespace impeller {
+
+//------------------------------------------------------------------------------
+/// @brief      Manages a queue of jobs that execute on a concurrent task
+///             runner. Jobs execute as soon as resources are available. Callers
+///             have the ability to re-prioritize jobs or perform jobs eagerly
+///             on their own threads if needed.
+///
+///             The job queue and all its methods are thread-safe.
+///
+class JobQueue final : public std::enable_shared_from_this<JobQueue> {
+ public:
+  //----------------------------------------------------------------------------
+  /// @brief      Creates a job queue which schedules tasks on the given
+  ///             concurrent task runner.
+  ///
+  /// @param[in]  task_runner  The task runner
+  ///
+  /// @return     A job queue if one can be created.
+  ///
+  static std::shared_ptr<JobQueue> Make(
+      std::shared_ptr<fml::ConcurrentTaskRunner> task_runner);
+
+  //----------------------------------------------------------------------------
+  /// @brief      Destroys the job queue.
+  ///
+  ~JobQueue();
+
+  JobQueue(const JobQueue&) = delete;
+
+  JobQueue& operator=(const JobQueue&) = delete;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Adds a job to the job queue and schedules it for execution at
+  ///             a later point in time. The job ID obtained may be used to
+  ///             re-prioritize the job within the queue at a later point.
+  ///
+  /// @param[in]  job   The job
+  ///
+  /// @return     The unique id for the job.
+  ///
+  UniqueID AddJob(fml::closure job);
+
+  //----------------------------------------------------------------------------
+  /// @brief      Prioritize a previously added job for immediate execution on
+  ///             the concurrent task runner.
+  ///
+  /// @param[in]  id    The job identifier.
+  ///
+  void PrioritizeJob(UniqueID id);
```

Comment:
Shouldn't this be used somewhere?  My understanding is that every pipeline would be scheduled to be created, then there is a mechanism to boost the priority when the pipeline is requested?

Reply:
Yes! I was going to add this the spot where the glyph atlas pipeline was being setup. Still WIP.

Reply:
Every pipeline will be scheduled eventually in the order in which the jobs were added. Jobs can skip to the front of the queue as necessary (this call). Jobs can be performed eagerly on anything too to avoid a context switch.

Reply:
Why do we need to re-prioritize the glyph atlas shader? Can we simply move it to the front of the queue in ContentContext?

Reply:
I think thats what I mean? Just call PrioritizeJob on it immediately after?

Reply:
What I mean is, we already process the jobs in FIFO order based on the initialization order in content_context.cc. So rather than adding a system for reprioritizing jobs, why don't we move the PSO in question earlier in the initialization setup?

Reply:
Example:

```diff
diff --git a/impeller/entity/contents/content_context.cc b/impeller/entity/contents/content_context.cc
index d1975f9f3f..9a352d05f8 100644
--- a/impeller/entity/contents/content_context.cc
+++ b/impeller/entity/contents/content_context.cc
@@ -296,6 +296,11 @@ ContentContext::ContentContext(
     solid_fill_pipelines_.CreateDefault(*context_, options);
     texture_pipelines_.CreateDefault(*context_, options);
     fast_gradient_pipelines_.CreateDefault(*context_, options);
+    glyph_atlas_pipelines_.CreateDefault(
+        *context_, options,
+        {static_cast<Scalar>(
+            GetContext()->GetCapabilities()->GetDefaultGlyphAtlasFormat() ==
+            PixelFormat::kA8UNormInt)});
 
     if (context_->GetCapabilities()->SupportsSSBO()) {
       linear_gradient_ssbo_fill_pipelines_.CreateDefault(*context_, options);
@@ -443,11 +448,6 @@ ContentContext::ContentContext(
                                                  options_trianglestrip);
   srgb_to_linear_filter_pipelines_.CreateDefault(*context_,
                                                  options_trianglestrip);
-  glyph_atlas_pipelines_.CreateDefault(
-      *context_, options,
-      {static_cast<Scalar>(
-          GetContext()->GetCapabilities()->GetDefaultGlyphAtlasFormat() ==
-          PixelFormat::kA8UNormInt)});
   yuv_to_rgb_filter_pipelines_.CreateDefault(*context_, options_trianglestrip);
   porter_duff_blend_pipelines_.CreateDefault(*context_, options_trianglestrip,
                                              {supports_decal});
```

Reply:
Oh right. Totally. I just didn't want to keep playing whack-a-mole for cases like these. Aaron had already done one pass at reordering pipelines https://github.com/flutter/engine/pull/52231. And there may be cases like these with different pipelines on other apps. The glyph atlas just happened to the one in this case.

TBH, all of this feels really icky because it can introduce jank past the frame. I'd rather these old devices just wait for the first frame content context setup. And I mean even delay the UI thread workloads so animations and such begin after setup. So that once it goes, its guaranteed to be smooth. But I'm just reducing alarm at the benchmark regression. I barely care about this device. Its massively internally synchronizing in a manner I haven't seen other devices do. And the other devices finish content context setup well in time.

If the preference is to just move the pipeline around and perhaps keep the addition of the pipeline name to the trace, that works too.

Reply:
I think that needing to add a PrioritizeJob call still counts as whack a mole in my mind. Adding the pipeline name to the timeline traces + re-arranging SGTM

Reply:
Roger, will re-arrange the pipeline and remove PrioritizeJob. Keep the jobqueue still or pass?

Reply:
The stall is completely gone (at least in this app) after the rearrange.
<img width="2080" alt="Screenshot 2024-08-06 at 11 52 55 AM" src="https://github.com/user-attachments/assets/dad209b4-379b-465a-8347-7a9c481742f4">

Reply:
I don't think we'll need the job queue

Reply:
Yeah, I think so too. Hopefully. Just checking the gallery and if I can suppress the regression on the mokey, that should be it.

Reply:
Closing in favor of https://github.com/flutter/engine/pull/54373